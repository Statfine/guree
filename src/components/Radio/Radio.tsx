import React, { ReactNode, useState, ReactEventHandler } from "react";
import "./Radio.scss";
import {
  combineClasses,
  scopedClass,
  classesObj,
  updateStateOnPropChange,
} from "@utils/index";

export interface radioAttrs {
  children: ReactNode;
  value: any;
  disabled: boolean;
  status: boolean;
}

export interface Props {
  children: ReactNode;
  value: any;
  disabled: boolean;
  status: boolean;
  onClick: (radioAttrs: radioAttrs) => {};
}

const Radio: React.FunctionComponent<Props> = ({
  children,
  value,
  disabled,
  status,
  onClick,
}: Props) => {
  const radioClass = (...classes: (string | Array<string> | classesObj)[]) =>
    scopedClass("radio", ...classes);

  const [selectStatus, setSelectStatus] = useState(status);
  const select = () => {
    setSelectStatus(true);
  };

  const radioClassName = combineClasses({
    [radioClass()]: true,
    [radioClass("selected")]: selectStatus,
    [radioClass("disabled")]: disabled,
  });
  const clickHandler: ReactEventHandler<HTMLElement> = (e) => {
    if (disabled) {
      return;
    }
    select();
    onClick && onClick({ children, value, disabled, status });
  };
  updateStateOnPropChange(status, setSelectStatus);
  return (
    <span className={radioClassName} onClick={clickHandler}>
      <span className={radioClass("inner")}></span>
      <span className={radioClass("name")}>{children}</span>
    </span>
  );
};

export default Radio;
