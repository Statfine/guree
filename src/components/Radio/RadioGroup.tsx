import React, { useState, ReactElement } from "react";
import { radioAttrs } from "./Radio";
import "./Radio.scss";
import { scopedClass, classesObj, updateStateOnPropChange } from "@utils/index";
import { isUndefined } from "util";

interface Props {
  children: ReactElement;
  value?: any;
  onChange?: (radioAttrs: radioAttrs) => {};
  defaultValue?: any;
}

const RadioGroup: React.FunctionComponent<Props> = ({
  children,
  value,
  onChange,
  defaultValue,
}: Props) => {
  const radioGroupClass = (
    ...classes: (string | Array<string> | classesObj)[]
  ) => scopedClass("radio-group", ...classes);

  const [checkedValue, setCheckedValue] = useState(
    isUndefined(value) ? defaultValue : value
  );

  const changeChecked = (radioAttrs: radioAttrs) => {
    const { value: checkedValue } = radioAttrs;
    React.Children.forEach(children, (element) => {
      const { value: radioValue } = element.props;
      if (radioValue === checkedValue) {
        onChange && onChange(checkedValue);
        if (!isUndefined(value)) {
          return;
        }
        setCheckedValue(radioValue);
      }
    });
  };

  updateStateOnPropChange(value, setCheckedValue);

  const radioChildren = React.Children.map(children, (element) => {
    return React.cloneElement(
      element,
      Object.assign(
        {},
        element.props,
        {
          checked: element.props.value === checkedValue,
          onChange: changeChecked,
        }
      )
    );
  });

  return <div className={radioGroupClass()}>{radioChildren}</div>;
};

export default RadioGroup;
