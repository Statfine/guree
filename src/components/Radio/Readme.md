### 基本用法

```jsx
<Radio value={"radio"}>radio</Radio>
```

### 选中状态

```jsx
<Radio status>radio</Radio>
```

### 禁用选项

```jsx
<Radio status={"radio"} disabled>radio</Radio>
```

### 单选组合
使用 RadioGroup 组件包裹多个 Radio 组件，可形成单选组合，通过 onChange 获取当前选中的值。注意: Radio 的 value 不能相同
```jsx
import { RadioGroup } from "@components/Radio";
<RadioGroup
  onChange={(radioAttrs) => {
    console.log(radioAttrs);
  }}
>
  <Radio value={"radio1"}>radio1</Radio>
  <Radio value={"radio2"}>radio2</Radio>
  <Radio value={"radio3"}>radio3</Radio>
</RadioGroup>;
```

### 控制选中
通过 value 控制当前选中的值
```jsx
import { RadioGroup } from "@components/Radio";
<RadioGroup value={"radio2"}>
  <Radio value={"radio1"}>radio1</Radio>
  <Radio value={"radio2"}>radio2</Radio>
  <Radio value={"radio3"}>radio3</Radio>
</RadioGroup>;
```
