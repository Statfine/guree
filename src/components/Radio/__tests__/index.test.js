import React from "react";
import renderer from "react-test-renderer";
import Radio, { RadioGroup } from "..";
import { mount, shallow } from "enzyme";

describe("Radio 组件", () => {
  it("正确导出 Radio 组件", () => {
    const radio = renderer.create(<Radio></Radio>).toJSON();
    expect(radio).toMatchSnapshot();
  });

  it("测试 Radio 组件 value 参数", () => {
    const radio = renderer.create(<Radio value={"radio"}></Radio>).toJSON();
    expect(radio).toMatchSnapshot();
  });

  it("测试 Radio 组件 status 参数", () => {
    const radio = renderer.create(<Radio status></Radio>).toJSON();
    expect(radio).toMatchSnapshot();
  });

  it("测试 Radio 组件 disabled 参数", () => {
    const radio = renderer.create(<Radio disabled></Radio>).toJSON();
    expect(radio).toMatchSnapshot();
  });

  it("测试 Radio 组件 onClick 参数", () => {
    const fn = jest.fn();
    const radio = mount(<Radio onClick={fn}></Radio>);
    radio.simulate("click");
    radio.update();
    expect(fn).toBeCalled();
    expect(radio.find(".guree-radio").hasClass("guree-radio-selected")).toEqual(
      true
    );
  });

  it("测试 Radio 组件 disabled 状态下的点击事件", () => {
    const fn = jest.fn();
    const radio = mount(<Radio onClick={fn} disabled></Radio>);
    radio.simulate("click");
    radio.update();
    expect(radio.find(".guree-radio").hasClass("guree-radio-selected")).toEqual(
      false
    );
  });
});

describe("RadioGroup 组件", () => {
  it("正确导出 RadioGroup 组件", () => {
    const radioGroup = renderer.create(<RadioGroup></RadioGroup>).toJSON();
    expect(radioGroup).toMatchSnapshot();
  });

  it("正确导出 RadioGroup 组件 value 参数", () => {
    const radioGroup = renderer
      .create(
        <RadioGroup value={"radio2"}>
          <Radio value={"radio1"}>radio1</Radio>
          <Radio value={"radio2"}>radio2</Radio>
          <Radio value={"radio3"}>radio3</Radio>
        </RadioGroup>
      )
      .toJSON();
    expect(radioGroup).toMatchSnapshot();
  });

  it("正确导出 RadioGroup 组件 onChange 参数", () => {
    const fn = jest.fn();
    const radioGroup = mount(
        <RadioGroup value={"radio2"} onChange={fn}>
          <Radio value={"radio1"}>radio1</Radio>
          <Radio value={"radio2"}>radio2</Radio>
          <Radio value={"radio3"}>radio3</Radio>
        </RadioGroup>
      );
    radioGroup.find(".guree-radio").forEach((node, index)=> {
      if (index === 2) {
        node.simulate("click")
      }
    })
    expect(fn).toBeCalled()
    radioGroup.update();
    radioGroup.find(".guree-radio").forEach((node, index)=> {
      expect(node.hasClass("guree-radio-selected")).toEqual(index === 2)
    })
  });
});
