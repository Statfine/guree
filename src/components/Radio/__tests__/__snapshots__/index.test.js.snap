// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio 组件 正确导出 Radio 组件 1`] = `
<span
  className="guree-radio"
  onClick={[Function]}
>
  <span
    className="guree-radio-inner"
  />
  <span
    className="guree-radio-name"
  />
</span>
`;

exports[`Radio 组件 测试 Radio 组件 disabled 参数 1`] = `
<span
  className="guree-radio guree-radio-disabled"
  onClick={[Function]}
>
  <span
    className="guree-radio-inner"
  />
  <span
    className="guree-radio-name"
  />
</span>
`;

exports[`Radio 组件 测试 Radio 组件 status 参数 1`] = `
<span
  className="guree-radio guree-radio-selected"
  onClick={[Function]}
>
  <span
    className="guree-radio-inner"
  />
  <span
    className="guree-radio-name"
  />
</span>
`;

exports[`Radio 组件 测试 Radio 组件 value 参数 1`] = `
<span
  className="guree-radio"
  onClick={[Function]}
>
  <span
    className="guree-radio-inner"
  />
  <span
    className="guree-radio-name"
  />
</span>
`;

exports[`RadioGroup 组件 正确导出 RadioGroup 组件 1`] = `
<div
  className="guree-radio-group"
/>
`;

exports[`RadioGroup 组件 正确导出 RadioGroup 组件 value 参数 1`] = `
<div
  className="guree-radio-group"
>
  <span
    className="guree-radio"
    onClick={[Function]}
  >
    <span
      className="guree-radio-inner"
    />
    <span
      className="guree-radio-name"
    >
      radio1
    </span>
  </span>
  <span
    className="guree-radio guree-radio-selected"
    onClick={[Function]}
  >
    <span
      className="guree-radio-inner"
    />
    <span
      className="guree-radio-name"
    >
      radio2
    </span>
  </span>
  <span
    className="guree-radio"
    onClick={[Function]}
  >
    <span
      className="guree-radio-inner"
    />
    <span
      className="guree-radio-name"
    >
      radio3
    </span>
  </span>
</div>
`;
